src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```
@@ -212,83 +156,124 @@ private UnfilteredPartitionIterator executeReadCommand(ReadCommand cmd, InetAddr
+                    }
 
-                @Override
-                public void onMergedRangeTombstoneMarkers(RangeTombstoneMarker merged, RangeTombstoneMarker[] versions)
-                {
-                    // cache the marker versions to be able to regenerate the original row iterator
-                    for (int i = 0; i < versions.length; i++)
-                        builders[i].addRangeTombstoneMarker(versions[i]);
-                }
+                    @Override
+                    public void onMergedRangeTombstoneMarkers(RangeTombstoneMarker merged, RangeTombstoneMarker[] versions)
+                    {
+                        // cache the marker versions to be able to regenerate the original row iterator
+                        for (int i = 0; i < versions.length; i++)
+                            builders[i].addRangeTombstoneMarker(versions[i]);
+                    }
 
-                @Override
-                public void close()
-                {
-                    for (int i = 0; i < sources.length; i++)
-                        originalPartitions.get(i).add(builders[i]);
-                }
-            };
+                    @Override
+                    public void close()
+                    {
+                        for (int i = 0; i < sources.length; i++)
+                            originalPartitions.get(i).add(builders[i]);
+                    }
+                };
+            }
         };
     }
 
+    private void incrementCachedRows()
+    {
+        currentRowsCached++;
+        
+        if (currentRowsCached == cachedRowsFailThreshold + 1)
+        {
+            String message = String.format("Replica filtering protection has cached over %d rows during query %s. " +
+                                           "(See 'cached_replica_rows_fail_threshold' in cassandra.yaml.)",
+                                           cachedRowsFailThreshold, command.toCQLString());
+
+            oneMinuteLogger.error(message);
```

Wondering about suppressing failure messages using the `NoSpamLogger`. Thinking about cases where enough requests hit the warning threshold but only a few hit the error threshold. Do we want warnings to suppress those errors?

---

@jrwest Yeah, I went back and forth on this a bit. It might be that there isn't much risk in allowing failure message to hit the regular log, given the system might simply not be able to process many queries like that concurrently anyway. I'm okay with changing this, assuming @adelapena has no objections.

---

Makes sense to me